### PR TITLE
Async Context Manager

### PR DESCRIPTION
**Pull request by t2t2**
Adds custom context manager that keeps parent through common async operations (eg. setTimeout, setImmediate (IE), Promise.then, ...). Patched targets are based on what's used in react ([v17](https://github.com/facebook/react/blob/v17.0.2/packages/scheduler/src/forks/SchedulerHostConfig.default.js), [newer](https://github.com/facebook/react/blob/8ea11306ad473b26a2c899ef7a893d25413f3510/packages/scheduler/src/forks/SchedulerDOM.js#L554)) and vue ([v2](https://github.com/vuejs/vue/blob/0603ff695d2f41286239298210113cbe2b209e28/src/core/util/next-tick.js#L42), [v3](https://github.com/vuejs/vue-next/blob/cd337149353d03446850b5f45b817f3fa160e5ff/packages/runtime-core/src/scheduler.ts#L107)), but should pretty much cover most of the things users do.

APMI-1716

## Using todolist example (react)

Current version:

image

image
image

Separate traceid for /items which happens after rerender caused by processing response of /login

With these changes:

image

image

Both requests are in same traceid

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -301,7 +302,9 @@ const SplunkRum: SplunkOtelWebType = {
       provider.addSpanProcessor(new SimpleSpanProcessor(new ConsoleSpanExporter()));
     }
 
-    provider.register();
+    provider.register({
+      contextManager: new SplunkContextManager()
```
Comment (jtmal-signalfx) [inline src/index.ts:306]: Opt in in the beginning is probably better.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,7 @@
 # Changelog
 
 ## Unreleased
+- Parent span propagation through async activities
```
Comment (jtmalinowski) [inline CHANGELOG.md:4]: > New `SplunkContextManager` for limited causality in Promise-based, React, and Vue flows
Something like this? I think propagation means linkage between services

File: integration-tests/tests/context/messageport-onmessage.ejs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+  <%- renderAgent({}) %>
+</head>
+<body>
+  <h1>MessagePort - onmessage</h1>
+
+  <button type="button" id="btn1">Button</button>
+
+  <script>
+    // Based on react/scheduler's implementation of nextTick
```
[the remark this answers is not in the document]
Comment (mhennoch) [reply integration-tests/tests/context/messageport-onmessage.ejs:15]: There is an issue in jira to have framework tests

File: src/SplunkContextManager.ts [hunk 4]
```diff
@@ -0,0 +1,571 @@
...
+    }
+
+    return listenersByType.get(type);
+  }
+
+  protected _patchEvents(): void {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const manager = this;
+
+    wrapNatively(XMLHttpRequest.prototype, 'addEventListener', original =>
```
[the remark this answers is not in the document]
Comment (t2t2) [reply src/SplunkContextManager.ts:359]: Would rather keep everything related to async trace connecting in one place (also made it way easier to config)

File: src/SplunkContextManager.ts [hunk 5]
```diff
@@ -0,0 +1,571 @@
...
+    const manager = this;
+
+    wrapNatively(XMLHttpRequest.prototype, 'addEventListener', original => 
+      function (...args: Parameters<XMLHttpRequest['addEventListener']>) {
+        if (isFunction(args[1])) {
+          const wrappedListeners = manager._getListenersMap(this, args[0]);
+          let wrapped = wrappedListeners.get(args[1] as EventListener);
+
+          if (!wrapped) {
+            wrapped = manager.bind(args[1] as EventListener, manager.active());
```
[the remark this answers is not in the document]
Comment (t2t2) [reply src/SplunkContextManager.ts]: Taking parent context as of registration is intentional, based on these thought processes in node side:

https://github.com/open-telemetry/opentelemetry-js/issues/1479
https://github.com/open-telemetry/opentelemetry-js/pull/2037

**Review by mhennoch — APPROVED** [review]

**Pull request merged**